python/test/test_storage_format.py

```diff
@@ -0,0 +1,273 @@
+            latest_part_header = part_header
+
+    with open(latest_part_header) as f:
+        part_header_json = json.load(f)
+        return part_header_json.get("kg.v1.storage_format_version")
+
+
+def flatten(dictionary, parent_key=False, separator="::", verbose=False):
+    """
+    Turn a nested dictionary into a flattened dictionary,
+    while also removing the random appended string from
+    the dictionary value
+    :param dictionary: The dictionary to flatten
+    :param parent_key: The string to prepend to dictionary's keys
+    :param separator: The string used to separate flattened keys
+    :return: A flattened dictionary
+    """
+
+    items = []
+    for key, value in dictionary.items():
+        if verbose:
+            print("checking:", key)
+        new_key = str(parent_key) + separator + key if parent_key else key
+        if isinstance(value, MutableMapping):
+            if verbose:
+                print(new_key, ": dict found")
+            if not value.items():
+                if verbose:
+                    print(f"Adding key-value pair: {new_key}, {None}")
+                items.append((new_key, None))
+            else:
+                items.extend(flatten(value, new_key, separator, verbose).items())
+        elif isinstance(value, list):
+            if verbose:
+                print(new_key, ": list found")
+            if len(value):
+                for k, v in enumerate(value):
+                    items.extend(flatten({str(k): v}, new_key, separator, verbose).items())
+            else:
+                if verbose:
+                    print(f"Adding key-value pair: {new_key}, {None}")
+                items.append((new_key, None))
+        else:
+            # remove the random appended string so we can actually compare two json files deterministically
+            if isinstance(value, str):
+                value = remove_rand_string(value)
+            if verbose:
+                print(f"Adding key-value pair: {new_key}, {value}")
+            items.append((new_key, value))
+    return dict(items)
+
+
+def matching_json(orig_file, other_file, verbose=False):
+    """
+    returns True if the structure and values of the two json files matches
+    """
+    with open(orig_file) as orig_f:
+        with open(other_file) as other_f:
+            orig = json.load(orig_f)
+            other = json.load(other_f)
+
+            if len(orig.keys()) != len(other.keys()):
+                print(
+                    f"{orig_file} json key number does not match {other_file},"
+                    f"diff = {list_diff(orig.keys(), other.keys())}"
+                )
+                return False
+
+            # flatten the json files.
+            # this also remove the random strings appended to some of the json value strings
+            # to make comparing the two possible
+            orig_flat = flatten(orig, verbose)
+            other_flat = flatten(other, verbose)
```

Comment:
I was wondering about this but do you have to flatten them? Can you not check for equivalence using `==`, which I thought was recursive...but perhaps not?

Reply:
hm, you do have a point. 
I was originally flattening them as a way to do comparison before I was also sanitizing out the random file names. 
I could probably get away without flattening them as well now, but would still have to traverse the entire dict to sanitize out the random strings. Let me think on this.

Reply:
so I've decided I like flattening them because it allows for explicitly calling out what is different:
```
            subset_flat = flatten(subset, verbose)
            superset_flat = flatten(superset, verbose)

            for val in subset_flat:
                if val not in superset_flat:
                    print(f"val = {val} not found in superset json file")
```
When the test fails, it will tell you exactly what value at what keys is different. 

We could say only flatten them when `==` fails so that we can print this out, but we still have to traverse the entire dict to sanitize all of the file names random strings so that wouldn't actually save us any time. 

I'll add a TODO to clean this up if/when we stop putting random strings in our file names.

Reply:
Sounds good - verbosity would help in situations like this. Unclear error messages are the worst.